<a href="https://colab.research.google.com/github/Twals20/Data-mining-/blob/main/DM_ProjectGroup_19.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Business Understanding
## 1.1 Background
The  Zambina legal system produces different types of legislative documents, such as *Bills*, *Acts*, and *Statutory Instruments (SIs)*. These documents are essential for governance, law enforcement, and civic education. However, due to the large volume of documents and their textual similarities, it is difficult for the general public and researchers to quickly identify and categorize them.

Manually classifying these documents is time-consuming and prone to errors. Therefore, applying data mining and machine learning techniques to automatically classify legislation can improve efficiency, accessibility, and public engagement with legal information.

## 1.2 Project Problem Statement
The increasing volume of Zambian legislative documents (Bills, Acts, and Statutory Instruments) makes manual classification inefficient, error-prone, and unsustainable. Currently, no automated system exists to categorize these documents, which limits accessibility and slows down research. This project addresses the problem by applying data mining and machine learning techniques to automatically classify legislative documents into their correct categories with improved accuracy and efficiency.

## 1.3 Data Mining Goals  

To achieve the business objectives, our project will focus on building a text classification model for legislative documents. Specifically, our goals are:  

1. Model Development: Train a supervised machine learning model (e.g., Logistic Regression, Random Forest, or Neural Network) that can classify legislative documents into predefined legal categories.  
2. Feature Extraction: Use Natural Language Processing (NLP) techniques such as TF-IDF or word embeddings to identify the most relevant keywords and features that drive the classification.  
3. Performance Evaluation: Validate the model using accuracy, precision, recall, and F1-score to ensure it meets the success criteria (≥ 80% accuracy).  
4. Interpretability: Ensure the model highlights the top features/keywords influencing each classification, making the results understandable for legal experts.  
5. Efficiency: Optimize the pipeline so that classification of a new document can be done in under 2 seconds.  

## 1.4 Project Objectives

1. Collect and Prepare Data:
Gather legislative documents (Bills, Acts, Statutory Instruments) from the National Assembly of Zambia and the Zambia Legal Information Institute, clean and label them for use in training and evaluation.

2. Develop a Classification Model:
Apply supervised machine learning techniques, using NLP-based feature extraction (e.g., TF-IDF, word embeddings), to classify documents into their correct categories.

3. Evaluate and Optimize Performance:
Validate the model using accuracy, precision, recall, and F1-score, aiming for at least 80% accuracy, and optimize classification speed to under 2 seconds per document.

4. Ensure Interpretability:
Incorporate mechanisms to highlight keywords or features influencing each classification, making outputs understandable to legal experts and non-experts.

5. Deliver a Usable System:
Package the trained model into a functional pipeline or prototype that can accept a legislative document and return its predicted category in real time.


## 1.5 Project Success Criteria

- *Model Accuracy: Achieve at least 80% classification accuracy* on a test dataset of legislative documents.
- *Interpretability*: The model should provide clear reasoning (e.g., top keywords or features) for each classification.
- *Practicality: The classification results should align with expert legal categorization at least 8 out of 10 times* when manually reviewed.
- *Efficiency: Classify a new document in under 2 seconds*.

- A working dataset of *at least 100–200 legislative documents* is collected and labeled.  
- The *project deliverables* (codebase, report, and presentation) are completed and submitted by the deadline.



In [ ]:
df.describe(include=['object'])


Title  Instrument-type   \
count                                                 112              112   
unique                                                110                5   
top     The Zambia Institute of Advanced Legal Educati...              Act   
freq                                                    2               63   

                     Citation  Year      Category   \
count                      109   110           112   
unique                     109    43            38   
top     Act number 23 of 2011   2021  Taxation Law   
freq                         1    10            30   

                                                 Summary   
count                                                 112  
unique                                                112  
top     An act to regulate the provision of accessible...  
freq                                                    1

# Data Understanding


## Dataset Summary

- **Number of rows:** 112  
- **Number of columns:** 6  

- **Columns and data types:**
  - Title: object (text)  
  - Instrument-type: object (text)  
  - Citation: object (text)  
  - Year: object (text)  
  - Category: object (text)  
  - Summary: object (text)  

- **Missing values:**
  - Title: 0  
  - Instrument-type: 0  
  - Citation: 3  
  - Year: 2  
  - Category: 0  
  - Summary: 0  

- **Numerical summary:**  
  - Year: Although stored as text, the most common year is 2021, with 10 occurrences.  

- **Categorical summary:**  
  - Title: 110 unique values, most common = "The Zambia Institute of Advanced Legal Education..." (appears 2 times)  
  - Instrument-type: 5 unique values, most common = "Act" (appears 63 times)  
  - Citation: 109 unique values, most common = "Act number 23 of 2011" (appears 1 time)  
  - Category: 38 unique values, most common = "Taxation Law" (appears 30 times)  
  - Summary: 112 unique values, most common = "An act to regulate the provision of accessible..." (appears 1 time)  

- **Observations:**  
  - The dataset is mostly complete; only the Citation and Year columns have a few missing values.  
  - Most legislation entries are of type "Act."  
  - Titles are mostly unique, indicating each row likely represents a different legal document.  
  - Category distribution shows some categories, like "Taxation Law," are more frequent than others.


In [ ]:
import pandas as pd
from google.colab import files
uploaded = files.upload()

df1 = pd.read_csv("zambia_legislation_dataset.csv")

Saving zambia_legislation_dataset.csv to zambia_legislation_dataset.csv


In [ ]:
import pandas as pd


df = pd.read_csv("/content/zambia_legislation_dataset.csv")

# Quick check
df.head()
df.shape
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112 entries, 0 to 111
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Title             112 non-null    object
 1   Instrument-type   112 non-null    object
 2   Citation          109 non-null    object
 3   Year              110 non-null    object
 4   Category          112 non-null    object
 5   Summary           112 non-null    object
dtypes: object(6)
memory usage: 5.4+ KB


In [ ]:
import pandas as pd

# Load your uploaded CSV
df = pd.read_csv("/content/zambia_legislation_dataset.csv")

# Check column names
print(df.columns)


Index(['Title ', 'Instrument-type ', 'Citation ', 'Year ', 'Category ',
       'Summary '],
      dtype='object')


In [ ]:
print(df.columns)


Index(['Title ', 'Instrument-type ', 'Citation ', 'Year ', 'Category ',
       'Summary '],
      dtype='object')


In [ ]:
df.columns = df.columns.str.strip()


In [ ]:
# Numeric summary
df.describe()            # Summary stats for numeric columns

# Year column exploration
df['Year'].value_counts()  # Count of laws per year

# Check for missing values in numeric columns
df[['Year']].isnull().sum()


Year    2
dtype: int64

In [ ]:
from google.colab import files
uploaded = files.upload()  # This opens a file picker to upload


Saving zambia_legislation_dataset.csv to zambia_legislation_dataset.csv


In [ ]:
#category counts
df['Category '].value_counts()

Category 
Taxation Law                                                         30
Health Law                                                           25
Education law                                                        16
Education law                                                         5
Financial Law, Legal Procedures                                       2
Agriculture, Plant Health, Trade Regulation                           2
Housing, Land Management, Urban Development                           1
Urban Development, Land Use, Governance                               1
Public Health, Agriculture, Pest Control                              1
Environmental Law, Pollution Control, Natural Resources               1
Wildlife Conservation, Environmental Law, Tourism                     1
Urban Planning, Land Use, Housing                                     1
Mining Law, Natural Resources, Economic Development                   1
Disaster Management, Public Safety, Governance                        1
Environmental Law, Pollution Control                                  1
Agricultural Law, Intellectual Property                               1
Tourism, Hospitality, Economic Development                            1
Environmental Law, Pollution Control, Natural Resource Management     1
Radiation Safety, Public Health                                       1
National Security, Anti-Terrorism, Financial Crime                    1
Mining Regulation, Governance, Economic Development                   1
Geological Survey, Mining, Natural Resources                          1
Appropriation Law, Finance, Governance                                1
Wildlife Conservation, Legal Procedure                                1
Climate Change, Environmental Law, Economic Policy                    1
Mining, Geological Survey, Economic Development                       1
Environmental Protection, Agriculture, Public Health                  1
Forestry, Governance                                                  1
Environmental Protection, Industrial Regulation                       1
Professional Regulation, Construction Law                             1
Environmental Protection, Water Management                            1
Environmental Protection, Climate Change, Trade Regulation            1
Wildlife Management, Tourism, Economic Regulation                     1
Local Government, Natural Resource Management, Conservation           1
Wildlife Conservation, Tourism, Ethical Hunting                       1
Wildlife Conservation, International Trade, CITES Compliance          1
Waste Management, Environmental Law, Industrial Regulation            1
Intellectual Property, Cultural Heritage, Nagoya Protocol             1
Name: count, dtype: int64

In [ ]:
#Instrument_type counts
df['Instrument-type '].value_counts()

Instrument-type 
Act                      63
Statutory Instrument     28
Bill                     11
Act                       6
Statutory instrument      4
Name: count, dtype: int64

In [ ]:
#unique values
df['Category '].unique()



array(['Education law', 'Education law ', 'Health Law',
       'Public Health,\xa0Agriculture,\xa0Pest Control',
       'Urban Development,\xa0Land Use,\xa0Governance',
       'Housing,\xa0Land Management,\xa0Urban Development',
       'Environmental Law,\xa0Pollution Control,\xa0Natural Resources',
       'Financial Law,\xa0Legal Procedures',
       'Mining Law,\xa0Natural Resources,\xa0Economic Development',
       'Disaster Management,\xa0Public Safety,\xa0Governance',
       'Wildlife Conservation,\xa0Environmental Law,\xa0Tourism',
       'Urban Planning,\xa0Land Use,\xa0Housing',
       'Mining Regulation,\xa0Governance,\xa0Economic Development',
       'Geological Survey,\xa0Mining,\xa0Natural Resources',
       'Environmental Law,\xa0Pollution Control',
       'Agricultural Law,\xa0Intellectual Property',
       'Tourism,\xa0Hospitality,\xa0Economic Development',
       'Environmental Law,\xa0Pollution Control,\xa0Natural Resource Management',
       'Radiation Safety,\xa0Publi